src/Sylius/Component/Core/Taxation/Applicator/OrderShipmentTaxesApplicator.php

```diff
@@ -61,29 +66,49 @@ public function apply(OrderInterface $order, ZoneInterface $zone): void
             return;
         }
 
-        $this->addAdjustment($order, (int) $taxAmount, $taxRate->getLabel(), $taxRate->isIncludedInPrice());
+        $this->addAdjustment($shipment, (int) $taxAmount, $taxRate, $shippingMethod);
     }
 
-    private function addAdjustment(OrderInterface $order, int $taxAmount, string $label, bool $included): void
-    {
-        /** @var AdjustmentInterface $shippingTaxAdjustment */
-        $shippingTaxAdjustment = $this->adjustmentFactory
-            ->createWithData(AdjustmentInterface::TAX_ADJUSTMENT, $label, $taxAmount, $included)
-        ;
-        $order->addAdjustment($shippingTaxAdjustment);
+    private function addAdjustment(
+        ShipmentInterface $shipment,
+        int $taxAmount,
+        TaxRateInterface $taxRate,
+        ShippingMethodInterface $shippingMethod
+    ): void {
+        $shipment->addAdjustment($this->adjustmentFactory->createWithData(
+            AdjustmentInterface::TAX_ADJUSTMENT,
+            $taxRate->getLabel(),
+            $taxAmount,
+            $taxRate->isIncludedInPrice(),
+            [
+                'shippingMethodCode' => $shippingMethod->getCode(),
+                'shippingMethodName' => $shippingMethod->getName(),
+                'taxRateCode' => $taxRate->getCode(),
+                'taxRateName' => $taxRate->getName(),
+                'taxRateAmount' => $taxRate->getAmount(),
+            ]
```

Comment:
If we would introduce detailed factories, this structure seems like the result of merging two previous factories.

In such aa situation, some recommendation for prefixes could be also a good idea.

Reply:
@lchrusciel 

Alternative solution,

```php
$adjustment->setDetailsFromValueObjects(
    new ShippingMethodVO($shippingMethod->getCode(), $shippingMethod->getName()),
    new TaxRateVO($taxRate->getCode(), $taxRate->getName(), $taxRate->getAmount())
);
```
Which will allow to specify new data more easily and offer extension point for developers.

For example, I've implemented `payment method charges` custom feature before and currently doing `country of origin` for Tax Rate.

But regardless of this two examples, I'm sure there are whole CS patterns on how to store/retrieve complex data in array field.

Reply:
for sure, what we've used is far from the ideal pattern, you are correct. We will think about alternatives. Hopefully, this details field will be good enough for now